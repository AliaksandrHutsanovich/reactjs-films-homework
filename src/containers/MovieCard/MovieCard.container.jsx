import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { REQUEST_TYPES } from '../../constants';
import { MovieCard } from '../../patterns';
import {
  useItemsSelector,
} from '../../hooks';

const MovieCardContainer = ({
  title,
  genresIds,
  rating,
  info,
  imgUrl,
  hoverable,
  renderHoverableContent,
  renderCoveringContent,
  id,
}) => {
  const [isShown, showCover] = useState(false);

  const genresById = useItemsSelector(REQUEST_TYPES.GENRES);

  return Object.keys(genresById).length ? (
    <MovieCard
      title={title}
      genres={genresIds.map((itemId) => (genresById[itemId].name))}
      rating={rating}
      info={info}
      imgUrl={imgUrl}
      hoverable={hoverable}
      renderHoverableContent={renderHoverableContent}
      renderCoveringContent={renderCoveringContent}
      showCoveringContent={isShown}
      onShowCover={showCover}
      id={id}
    />
  ) : null;
};

MovieCardContainer.propTypes = {
  title: PropTypes.string.isRequired,
  genresIds: PropTypes.arrayOf(
    PropTypes.number,
  ).isRequired,
  rating: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  imgUrl: PropTypes.string.isRequired,
  hoverable: PropTypes.bool,
  renderHoverableContent: PropTypes.func,
  renderCoveringContent: PropTypes.func,
  info: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
};

const renderDefaultContent = () => (<div />);

MovieCardContainer.defaultProps = {
  hoverable: false,
  renderHoverableContent: renderDefaultContent,
  renderCoveringContent: renderDefaultContent,
};

export default MovieCardContainer;
