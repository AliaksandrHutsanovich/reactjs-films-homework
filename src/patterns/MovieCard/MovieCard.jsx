import React, { memo } from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import { Card } from 'antd';
import { Image } from '../../containers';

import '!style-loader!css-loader!sass-loader!antd/dist/antd.css'; //eslint-disable-line

import MovieDescription from '../MovieDescription';
import preload from '../../assets/images/preload.png';

import styles from './MovieCard.scss';

const { Meta } = Card;

const MovieCard = ({
  title,
  genres,
  rating,
  imgUrl,
  info,
  hoverable,
  renderHoverableContent,
  renderCoveringContent,
  showCoveringContent,
  onShowCover,
  id,
}) => (
  <Card
    className={clsx(styles.card, { [styles.hoverable]: hoverable && !showCoveringContent })}
    cover={(
      <div className={styles.imageWrapper}>
        <Image preload={preload} alt="cardImg" imgUrl={imgUrl} className={styles.image} spinClassName={styles.imageSpin} />
        <div className={styles.hoverableContentWrapper}>
          {renderHoverableContent(onShowCover)}
        </div>
      </div>
    )}
  >
    <Meta
      description={(
        <MovieDescription
          title={title}
          genres={genres}
          rating={rating}
          ratingFontType="bordered"
          id={id}
        />
      )}
    />
    <div
      className={clsx(
        styles.coveringContentWrapper,
        { [styles.showCover]: showCoveringContent },
      )}
    >
      {renderCoveringContent(
        title,
        genres,
        rating,
        info,
        onShowCover,
        imgUrl,
      )}
    </div>
  </Card>
);

MovieCard.propTypes = {
  title: PropTypes.string.isRequired,
  genres: PropTypes.arrayOf(
    PropTypes.string,
  ).isRequired,
  rating: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  imgUrl: PropTypes.string.isRequired,
  hoverable: PropTypes.bool,
  renderHoverableContent: PropTypes.func,
  renderCoveringContent: PropTypes.func,
  info: PropTypes.string,
  showCoveringContent: PropTypes.bool,
  onShowCover: PropTypes.func,
  id: PropTypes.number.isRequired,
};

export const renderDefaultContent = () => (<div />);

MovieCard.defaultProps = {
  hoverable: false,
  renderHoverableContent: renderDefaultContent,
  renderCoveringContent: renderDefaultContent,
  showCoveringContent: false,
  onShowCover: null,
  info: '',
};

export default memo(MovieCard); // additional memoization is possible
