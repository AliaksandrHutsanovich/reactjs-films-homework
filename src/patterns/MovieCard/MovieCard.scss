@import '../../styles/global';

$imageWidth: 390;
$imageHeight: 220;

.card {
  width: #{$cardWidth}px;
  position: relative !important;
  height: #{$cardHeight}px;
  overflow: hidden;
  margin: 10px 0 !important;
  border-radius: 4px 4px 0 0 !important;
}

.imageWrapper {
  width: 100%;
  height: #{$imageHeight}px;
  position: relative;
}

.image {
  width: #{$imageWidth}px;
  margin: 0 -#{($imageWidth/2 - $cardWidth/2)}px;
  height: 100%;
}

.imageSpin {
  top: #{$imageHeight/3}px;
  left: #{$cardWidth/3}px
}

.hoverableContentWrapper {
  position: absolute;
  top: 0;
  left: 0;
  width: #{$cardWidth}px;
  height: #{$imageHeight}px;
  background-color: rgba(0, 0, 0, 0);
  display: none;
}

.hoverable {
  cursor: pointer;

  &:hover {
    .hoverableContentWrapper {
      display: block;
      transition: background-color 1s;

      &:hover {
        background-color: rgba(0, 0, 0, 0.5);
      }
    }
  }
}

.coveringContentWrapper {
  position: absolute;
  top: 0px;
  transform: translate(0, -#{$cardHeight + 20}px);
  left: 0;
  width: #{$cardWidth}px;
  height: #{$cardHeight}px;
  transition: transform 1s ease-in;
}

.showCover {
  transform: translate(0, 0);
}
