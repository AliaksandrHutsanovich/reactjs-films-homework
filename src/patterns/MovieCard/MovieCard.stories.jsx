import React from 'react';
import MovieCard from './MovieCard';
import image from '../../assets/images/story-background.png';

export default {
  component: MovieCard,
  title: 'patterns/MovieCard',
};

const genres = [
  'Action',
  'Adventure',
  'Fantasy',
];

export const cardExample = () => (
  <MovieCard
    title="ASSASIN'S CREED"
    genres={genres}
    rating="4.7"
    imgUrl={image}
    hoverable
    renderHoverableContent={() => <p>This is hoverable content</p>}
  />
);

export const coveringCardExample = () => (
  <MovieCard
    title="ASSASIN'S CREED"
    genres={genres}
    rating="4.7"
    imgUrl={image}
    hoverable
    renderHoverableContent={() => <p>This is hoverable content</p>}
    showCoveringContent
    renderCoveringContent={() => (
      <div style={{ width: '220px', height: '220px', backgroundColor: 'rgba(0, 0, 0, 0.7)' }}>
        <h2 style={{ color: 'white' }}>This is covering content</h2>
      </div>
    )}
  />
);
