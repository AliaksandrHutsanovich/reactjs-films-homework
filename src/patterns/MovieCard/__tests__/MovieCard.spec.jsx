import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import MovieCard, { renderDefaultContent } from '../MovieCard';
import { coveringCardExample } from '../MovieCard.stories';

const genres = [
  'Action',
  'Adventure',
  'Fantasy',
];

describe('<MovieCard />', () => {
  const renderer = new ShallowRenderer();

  it('card should be rendered properly', () => {
    renderer.render(
      <MovieCard
        id={3}
        title="ASSASIN'S CREED"
        genres={genres}
        rating="4.7"
        imgUrl=""
        hoverable
        renderHoverableContent={() => <p>This is hoverable content</p>}
        renderCoveringContent={renderDefaultContent}
      />,
    );

    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('card with covering content should be rendered properly', () => {
    renderer.render(coveringCardExample());
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
