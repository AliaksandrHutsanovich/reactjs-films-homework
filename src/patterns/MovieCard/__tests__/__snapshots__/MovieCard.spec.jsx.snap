// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieCard /> card should be rendered properly 1`] = `
<Card
  className="card hoverable"
  cover={
    <div
      className="imageWrapper"
    >
      <UNDEFINED
        alt="cardImg"
        className="image"
        imgUrl=""
        preload="test-file-stub"
        spinClassName="imageSpin"
      />
      <div
        className="hoverableContentWrapper"
      >
        <p>
          This is hoverable content
        </p>
      </div>
    </div>
  }
>
  <Meta
    description={
      <UNDEFINED
        genres={
          Array [
            "Action",
            "Adventure",
            "Fantasy",
          ]
        }
        id={3}
        rating="4.7"
        ratingFontType="bordered"
        title="ASSASIN'S CREED"
      />
    }
  />
  <div
    className="coveringContentWrapper"
  >
    <div />
  </div>
</Card>
`;

exports[`<MovieCard /> card with covering content should be rendered properly 1`] = `
<Card
  className="card"
  cover={
    <div
      className="imageWrapper"
    >
      <UNDEFINED
        alt="cardImg"
        className="image"
        imgUrl="test-file-stub"
        preload="test-file-stub"
        spinClassName="imageSpin"
      />
      <div
        className="hoverableContentWrapper"
      >
        <p>
          This is hoverable content
        </p>
      </div>
    </div>
  }
>
  <Meta
    description={
      <UNDEFINED
        genres={
          Array [
            "Action",
            "Adventure",
            "Fantasy",
          ]
        }
        id={2}
        rating="4.7"
        ratingFontType="bordered"
        title="ASSASIN'S CREED"
      />
    }
  />
  <div
    className="coveringContentWrapper showCover"
  >
    <div
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.7)",
          "height": "220px",
          "width": "220px",
        }
      }
    >
      <h2
        style={
          Object {
            "color": "white",
          }
        }
      >
        This is covering content
      </h2>
    </div>
  </div>
</Card>
`;
